### ISSUE TITLE
Introduce a InferValueTuples=X compiler flag.

### ISSUE THREAD

**Issue by OnurGumus** [fsharp/fslang-suggestions#813]
# Introduce a compiler switch as InferAsValueTuples = X

I propose we introduce a compiler switch to change the default inference of tuples as value tuples where InferAsValueTuples = X will cause all tuples with length less than X being inferred as ValueTuples.
The default value of X will be 0 and will effectively turn off this feature.

The existing way of approaching this problem in F# is to use struct(x,y) but many times the users don't use the struct keyword and this creates the following problems
- It requires more typing.
- It creates a false sense of correctness as in ref tuples are better defaults since arguably especially for smaller tuples, ValueTuples perform better and are better defaults.

C# people are already using ValueTuples by default.

## Pros and Cons 

The advantages of making this adjustment to F# are 
- People will use saner default as ValueTuples
- Less typing and easier reading
- Better C# ValueTuple interop
- At source code level it won't break anything
- Gradually increase the default value of X if needed.

The disadvantages of making this adjustment to F# are 
- It may cause confusing where each developer using a different values for X but source will still compile no matter what.
- The author of the code has to remember what X is, but still he will get a meaningful error message if he goes wrong.
-  If you really need a ref tuple in all cases you must give the type hint explicitly.
`let (x,y):(int*int) = 3,5`

## Extra information
S

Related suggestions: (put links to related suggestions here)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [X] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [X] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [X] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [X] This is not a breaking change to the F# language design
* [X] I or my company would be willing to help implement and/or test this

**Comment by cartermp** [comment]
> At source code level it won't break anything

This is not true. The following code will fail to compile if `x` is inferred to be a ValueTuple.

```fsharp
let x = (1, 2)
let y = x.ToValueTuple()
```

`X` would also have to be the size in bytes, not the number of elements in a tuple, making the ergonomics of this suggestion a bit more difficult.

**Comment by OnurGumus** [comment]
@cartermp that's not what I meant. You don't need that method, if you have any api consumes ref tuples it will break anyway :) What I meant was, since default would be zero, adding this flag won't break anything.

**Comment by Happypig375** [comment]
Before adding this, #782 would need to be addressed to make this feature more usable.

**Comment by cartermp** [comment]
@OnurGumus While it's true that one wouldn't _need_ that method, that doesn't change the fact that this suggestion is a breaking change. A valid F# program today that calls this method on a reference tuple will fail to compile if this were implemented in the next F# language version, thus making this a source-breaking change.

**Comment by OnurGumus** [comment]
@cartermp OK I moved it to disadvantages stating it may break things.

**Comment by varon** [comment]
Please can we get this in.

**Comment by abelbraaksma** [comment]
I'm not sure how this could possibly work. Value tuples and normal tuples have a disjunct set of functions that apply to them. Also, any member of a bcl class takes a reference tuple, as do table style match patterns. Changing this and making it work with our without this switch seems a huge undertaking, by no means 'S', or perhaps I misunderstand the suggestion.

That's not even mentioning the possible negative impact on performance, or running out of stack space (value tuples can take up significantly more stack space, iirc). Carefully crafting a balance between value tuples and reference tuples is at the heart of good performance tuning. A switch to turn reference tuples magically into value tuples is not gonna change that, if anything, it would make tuning more complex. 

As an exercise, try to code with `#if INFERVALUETUPLES` and write the different code paths by hand in a test project, to find out if it is at all feasible to automate this process (it's what a compiler would need to do). I think it isn't, but I'd happily be proven wrong.

**Comment by TIHan** [comment]
Adding a switch that can change language semantics like this is a slippery slope. One project could use ValueTuple where as another would be Tuple; harder to reason about between projects. It's better to stay consistent with what we have.

I understand a lot of folks do like ValueTuple for C# and wish F# could have that by default. However, the practice of using tuples in C# and F# are slightly different. In F#, you tend to pass Tuples around more often than C#. If F# defaulted to ValueTuple, you could be incurring extra copying and getting worse performance as a result. Also, F# can de-tuple expressions in some cases; therefore, no allocation of any kind.

**Comment by 7sharp9** [comment]
>ValueTuples perform better and are better defaults

This has proven to not be the case.  

Tuples are far less prevalent in C# so I can understand the default there but not in F#.

**Comment by varon** [comment]
Rather than absolute throughput, do note that some tasks and workloads are incredibly sensitive to garbage which forces us to write `struct` all over the place.

Avoiding this with clean syntax (even at a possible throughput cost in some edge cases) would be a great improvement.

**Comment by abelbraaksma** [comment]
How is `struct` not clean and clear? It conveys the intend and meaning and is only a few keystrokes long. Or you can use type aliases for more clarity and brevity.

But apart from explaining the status quo, which I'm sure you're aware of, assuming we would want this in, I don't see any technical way of getting it done apart from breaking code. And even if you'd allow a switch that changed behavior (which this does), how can the compiler possibly understand when you need the reference version of a tuple? Don't forget, as explained before, the whole language is designed around tuples. Would you suggest that if we lose `struct`, we introduce a keyword to mean 'reference tuple'?

Either way, so far I've not seen a suggestion to do this without breaking code and behavior, and without very significant changes to the language as a whole.

**Comment by cartermp** [comment]
> Rather than absolute throughput, do note that some tasks and workloads are incredibly sensitive to garbage which forces us to write `struct` all over the place.

We're well aware. But it's also worth noting that just because `struct` is one something, doesn't mean it won't generate lots of garbage.

`struct (Array.zeroCreate<byte> 85_000, Array.zeroCreate<byte> 85_000)` will allocate twice on the large object heap, which can lead to the GC to pause all threads as it cleans up data.

**Comment by Happypig375** [comment]
That will be 3 allocations wirh reference tuples compared to 2 allocations with struct tuples, won't it? One less allocation is still better, although not having a large effect on overall performance.

**Comment by abelbraaksma** [comment]
> That will be 3 allocations wirh reference tuples
Why? Do you mean that creates three large arrays? I don't think so. Both struct and reference tuples will create the arrays on the LO heap. I'd be surprised if there's a difference when using struct vs reference tuples where the contents contains references.

I may be wrong, but unless you use struct tuples with struct data, the speed gain or resources gain will be minimal.

**Comment by Happypig375** [comment]
@abelbraaksma 
>One less allocation is still better, although not having a large effect on overall performance.

🙂

**Comment by cartermp** [comment]
> That will be 3 allocations wirh reference tuples compared to 2 allocations with struct tuples, won't it? One less allocation is still better, although not having a large effect on overall performance.

Perf-wise it's noise, so it doesn't matter (assuming the compiler doesn't optimize away the tuple in the first place).

**Comment by dsyme** [comment]
There is another possibility here, which is to add an option to "default to struct tuples if nothing else constrains a tuple to be a ref tuple".

That is, if the tuple is purely internal to an assembly, by some measure of internal.

There are many problems with this - among them being that 
- the user may not understand the rules
- in scripts and projects without signature files little is typically internal
- there are cases that could give a perf loss

But it's perhaps something to consdier long term

**Comment by dsyme** [comment]
I would be very interested to see a community-led prototype of this, applied to a large F# compiler codebase such as the F# compiler, both to see how much code needs explicit ref-tuple annotation, and to see if there is any actual performance difference.

**Comment by TIHan** [comment]
There is also a possibility the JIT could optimize tuples. I will want to experiment with this idea in .NET 8.
